### PR TITLE
Added collapsible sidebar sections

### PR DESCRIPTION
**Pull request by mostlyvirtual**
This might help with #588 as well.

### CODE REVIEW

File: assets/components/MainMenu.svelte [hunk 1]
```diff
@@ -20,6 +20,8 @@ let navEl;
 let searchHasFocus = false;
 let visibleLinks = [];
 
+let collapsedStates = JSON.parse(localStorage.getItem('collapsedStates')) || {};
```
Comment (jhthorsen) [inline assets/components/MainMenu.svelte:23]: What happens to JSON.parse() when collapsedStates is not defined in localStorage?

File: assets/components/MainMenu.svelte [hunk 2]
```diff
@@ -152,9 +154,21 @@ function onSearchKeydown(e) {
...
+
+function toggleSection(connection_id) {
+    if (collapsedStates[connection_id]) {
+        collapsedStates[connection_id] = !collapsedStates[connection_id];
+    } else {
+        collapsedStates[connection_id] = true; // default to collapsed if not already in the object
+    }
+
+    // Save to localStorage
+    localStorage.setItem('collapsedStates', JSON.stringify(collapsedStates));
```
Comment (jhthorsen) [inline assets/components/MainMenu.svelte:166]: Please change the indentation for consistency.

File: assets/components/MainMenu.svelte [hunk 3]
```diff
@@ -152,9 +154,21 @@ function onSearchKeydown(e) {
 function renderUnread(conversation, max = 60) {
   return conversation.unread > max ? (max + '+') : conversation.unread || 0;
 }
+
+function toggleSection(connection_id) {
+    if (collapsedStates[connection_id]) {
+        collapsedStates[connection_id] = !collapsedStates[connection_id];
+    } else {
+        collapsedStates[connection_id] = true; // default to collapsed if not already in the object
+    }
```
Comment (jhthorsen) [inline assets/components/MainMenu.svelte:163]: Don't think you need all the if/else logic. Just do:

```javascript
function toggleSection(connection_id) {
  collapsedStates[connection_id] = !collapsedStates[connection_id];
  localStorage.setItem('collapsedStates', JSON.stringify(collapsedStates));
}
```

Or do you want the sections to be collapsed as default..? I'm not very fond of hiding information from new users.

File: assets/components/MainMenu.svelte [hunk 4]
```diff
@@ -152,9 +154,21 @@ function onSearchKeydown(e) {
...
+
+    // Save to localStorage
+    localStorage.setItem('collapsedStates', JSON.stringify(collapsedStates));
+}
+
 </script>
 
-<style>
+<style lang="scss">
```
Comment (jhthorsen) [inline assets/components/MainMenu.svelte:171]: Why is this changed?

File: assets/components/MainMenu.svelte [hunk 5]
```diff
@@ -184,80 +198,97 @@ function renderUnread(conversation, max = 60) {
...
+        <b class="badge-connection" hidden="{!connection.unread}" on:click={e => {
+          toggleSection(connection.connection_id);
+          }}>{renderUnread(connection)}
+        </b>
+        <!-- Here's an icon indicating collapsed/expanded state -->
+        <div class="chevron"><Icon
+          name={collapsedStates[connection.connection_id] ? 'chevron-right' : 'chevron-down'}
+          on:click={e => {
+          toggleSection(connection.connection_id);}}/>
+        </div>
```
Comment (jhthorsen) [inline assets/components/MainMenu.svelte:221]: This section has a bit strange formatting. Here is my suggestion:

```html
<b class="badge-connection" hidden="{!connection.unread}" on:click={() => toggleSection(connection.connection_id)}}>{renderUnread(connection)}</b>
<!-- Here's an icon indicating collapsed/expanded state -->
<div class="chevron">
  <Icon name={collapsedStates[connection.connection_id] ? 'chevron-right' : 'chevron-down'}
    on:click={e => toggleSection(connection.connection_id)}/>
</div>

File: assets/components/MainMenu.svelte [hunk 6]
```diff
@@ -184,80 +198,97 @@ function renderUnread(conversation, max = 60) {
...
-    {#if $user.is('admin')}
-      <Link href="/settings">
-        <Icon name="tools"/>
-        <span>{$l('Settings')}</span>
+
+    <h3 class="account-toggle-plus" on:click={() => toggleSection($user.email)}>
+      <span>{$user.email || $l('Account')}</span>
+      <Icon name={collapsedStates[$user.email] ? 'plus' : 'minus'} />
+    </h3>
+    <div class={collapsedStates[$user.email] ? 'collapsedSection' : ''}>
```
Comment (jhthorsen) [inline assets/components/MainMenu.svelte:244]: Not a fan of camelCase in classnames. But either way, I think you can just use the "hidden" class, which already exists. I would even suggest using that on the links above...

**Review by jhthorsen — CHANGES_REQUESTED** [review]
Some things that I think needs tuning, but overall I think we'll add the functionality.

Just one big thing: Right now, I don't think you're able to search for "hidden" conversations. I think it makes sense that the conversations appear if searched for.

**Comment by mostlyvirtual** [comment]
Thanks for taking the time to review this. I'll look into each requested change and reply in-line. Probably next week.

File: assets/components/MainMenu.svelte (hunk 1 above)
Comment (mostlyvirtual) [reply assets/components/MainMenu.svelte:23]: I think the || {} fallback should take care of that. If collapsedStates is not defined in localStorage, then it will be initialized as an empty object.

File: assets/components/MainMenu.svelte (hunk 3 above)
Comment (mostlyvirtual) [reply assets/components/MainMenu.svelte:163]: I think you're right. If there is no cookie set, then all sections should be expanded, else collapse the ones that need to be collapsed. I've updated the code with your suggestion.

File: assets/components/MainMenu.svelte (hunk 4 above)
Comment (mostlyvirtual) [reply assets/components/MainMenu.svelte:171]: To be honest, I think it might've been from some earlier testing and hacking. I've removed that change for now, and will run through the test cases to see if I can replicate the reason why I put it in in the first place.

File: assets/components/MainMenu.svelte (hunk 6 above)
Comment (mostlyvirtual) [reply assets/components/MainMenu.svelte:244]:
```suggestion
    <div class="collapsedSection" hidden={!collapsedStates[$user.email]}>
```

File: assets/components/MainMenu.svelte (hunk 5 above)
Comment (mostlyvirtual) [reply assets/components/MainMenu.svelte:221]:
```suggestion
        <b class="badge-connection" hidden="{!connection.unread}" on:click={e => { toggleSection(connection.connection_id); }} on:keydown={e => { if (e.keyCode === 13) { toggleSection(connection.connection_id); } }}>{renderUnread(connection)}</b>
        <!-- Here's an icon indicating collapsed/expanded state -->
        <div class="chevron">
          <Icon name={collapsedStates[connection.connection_id] ? 'chevron-right' : 'chevron-down'}
          on:click={e => { toggleSection(connection.connection_id);}}/>
        </div>
```

File: assets/components/MainMenu.svelte (hunk 6 above)
Comment (mostlyvirtual) [reply assets/components/MainMenu.svelte:244]: Actually, that's wrong, it should be:
    <div hidden={!collapsedStates[$user.email]}>

Comment (mostlyvirtual) [reply assets/components/MainMenu.svelte:244]:
```suggestion
    <div class={collapsedStates[$user.email] && !filter.length ? 'collapsed-section' : ''}>
```

I'm making this suggestion, as I tried using the hidden class, but then the filter/search for things like "Files" or "Settings" wouldn't work anymore.

**Comment by mostlyvirtual** [comment]
> Just one big thing: Right now, I don't think you're able to search for "hidden" conversations. I think it makes sense that the conversations appear if searched for.

Actually, this looks like it was an easy and clean fix after about 2 hours of trying ugly solutions :)

It involved adding filter.length to this line for the conversations/channels:
`      {#if !collapsedStates[connection.connection_id] || filter.length}
`

And filter.length to this line for the account section:

`    <div class={collapsedStates[$user.email] && !filter.length ? 'collapsedSection' : ''}>
`

File: assets/components/MainMenu.svelte (hunk 1 above)
Comment (jhthorsen) [reply assets/components/MainMenu.svelte:23]: I just tried it myself, and it seems like you're right. At least in Firefox. I'm not sure why though, since it doesn't look like parse() should be able to take null. Please change it just to be sure:

```javascript
let collapsedStates = JSON.parse(localStorage.getItem('collapsedStates') || '{}');
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse

Comment (mostlyvirtual) [reply assets/components/MainMenu.svelte:23]: I'm not entirely sure what the difference is tbh, but from what I can tell it also doesn't hurt, so I implemented your snippet.

**Review by marcusramberg — APPROVED** [review]
lgtm!

**Comment by jhthorsen** [comment]
So busy today -- forgot to say thanks 😄

**Comment by mostlyvirtual** [comment]
> So busy today -- forgot to say thanks 😄

My pleasure. I feel you on the busy part, don't worry. Glad we reached a point where we could merge. Thank you.
